src/cmd/sign.js

```diff
@@ -125,6 +156,170 @@ export default function sign(
+    }
+  }
+
+  // manifestData.version
+  if (manifestData.name == null) {
+    throw new WebExtError(
+      'update-link was passed but the manifest.json ' +
+      'is missing the .name property. ' +
+      'therefore, no updateManifest could be generated');
+  }
+
+  // ..lastly check if the passed updateLink parameter is correct
+  if (updateLink.indexOf('{xpiFileName}') === -1) {
+    throw new WebExtError(
+      'Unable to parse --update-link url, please use {xpiFileName} ' +
+      'as a substitute for the XPI file');
+  }
+
+  return true;
+}
+
+/**
+ * Generates a new updateManifest.json file
+ * It will add an entry to a previously fetched updateManifest.json file
+ * for each generated XPI file.
+ * The result will be stored in the artifacts directory
+ */
+// jslint incorrectly reports that this return statement should be a yield.
+// eslint-disable-next-line
+async function createNewUpdateManifest(
+  id: ?string, XPIPath: string, artifactsDir,
+  manifestData: Object, updateLink: string, oldUpdateManifestData: Object
+) {
+  let updateManifestFileName = '';
+  let newUpdateManifest;
+
+  let addonName = manifestData.name || 'your application';
+
+  let parsed = url.parse(manifestData.applications.gecko.update_url);
+  // these if's are really only here to make flow happy
+  // it doesn't seem to accept that path.basename returns a string
+  if (parsed != null && parsed.pathname != null) {
+    updateManifestFileName = path.basename(parsed.pathname);
+  }
+
+  let newVersion = {
+    version: manifestData.version,
+    update_link: updateLink.replace('{xpiFileName}', path.basename(XPIPath)),
+  };
+
+  if (oldUpdateManifestData.addons == null) {
+    // if the addons property did not exist,
+    // it means that the updateManifest is completely new.
+    // In case the developer forgot to add .addons, do it here automatically
+    oldUpdateManifestData.addons = {};
+  }
+
+  if (id && oldUpdateManifestData.addons[id]) {
+    // append a new release to the updateManifest
+    if (oldUpdateManifestData.addons[id].updates) {
+      oldUpdateManifestData.addons[id].updates.push(newVersion);
+    }
+  } else {
+    // the extensionID was not found in the .addons property
+    // this can mean either that this is the first release of an extension
+    // OR that the extensionID changed,
+    // we'll warn the user about the second scenario just to be sure.
+    log.warn(`Creating first release of ${addonName}`);
+    log.warn(
+      'If this is not the actual first release, check your extension-id');
+    oldUpdateManifestData.addons[id] = {};
+    oldUpdateManifestData.addons[id].updates = [newVersion];
+  }
+
+  // the oldUpdateManifest has been updated, it's ready to be saved
+  // prettify it so it's readable to the maintainers
+  newUpdateManifest = JSON.stringify(oldUpdateManifestData, null, 4);
+
+  return fs.writeFile(
+    path.join(artifactsDir, updateManifestFileName),
+    newUpdateManifest,
+    function(error) {
```

Comment:
you can't use the error callback here since you are returning a promise -- i.e. the caller needs be notified about the error too. You can use chaining instead:
````javascript
return fs.writeFile(...)
  .catch((error) => {
    throw new WebExtError(
      `Was unable to write updated ${addonName}: ${error}`);
  });
````